PrepareFiles.py

```diff
@@ -1,106 +1,164 @@
-                output_file_list.append(this_pdf)
+            if length == count:
+                this_file.append(file['name'])
+                output_file_list.append(this_file)
+            elif (file['size'] + data[count]['size']) > max_deliveryfile_size:
+                this_file.append(file['name'])
+                output_file_list.append(this_file)
                 size = 0
-                this_pdf = []
-            elif length == count:
-                output_file_list.append(this_pdf)
+                this_file = []
+            else:
+                this_file.append(file['name'])
+                size = file['size'] + size
+                if size > max_deliveryfile_size:
+                    output_file_list.append(this_file)
+                    size = 0
+                    this_file = []
             count += 1
         return output_file_list
 
-    def _create_pdf(self, max_deliveryfile_size, reference):
-        # TODO update digitalfile mata data with progress
-        # TODO keep track of list of PDFs
-        # TODO write list of PDFs back to digitalfile meta data
-        batch_list = self._create_image_list(max_deliveryfile_size, self.replica_data['files'])
+    def _create_zip(self, max_deliveryfile_size, reference, iaid):
+        print(self.replica_data['files'])
+        batch_list = self._create_file_list(max_deliveryfile_size, self.replica_data['files'])
+        zipfile_name_prefix = self._create_file_name_prefix(reference)
+        n = 1
+        start = self._post_progress(iaid, 1)
+        for batch in batch_list:
+            output_name = zipfile_name_prefix + '{:02d}'.format(n) + '.zip'
+            with zipfile.ZipFile('/tmp/' + output_name, 'w') as zip_file:
+                for file_key in batch:
+                    try:
+                        file_name = file_key.rpartition('/')[2]
+                        print(file_name)
+                        s3_client.download_file(s3_bucket_get, file_key, '/tmp/' + file_name)
+                        zip_file.write('/tmp/' + file_name)
+                        os.remove('/tmp/' + file_name)
+                    except ClientError as e:
+                        print(e)
+                        print('[ERROR 404 NoSuchKey] - Failed to retrieve file from s3: ' + file_key)
+                        break
+            zip_file.close()
+            success = self._s3_put_object(s3_bucket_put, '/tmp/' + output_name, 'test/' + output_name)
+            n += 1
+        return {"code": 200}
+
+    def _create_pdf(self, max_deliveryfile_size, reference, iaid):
+        batch_list = self._create_file_list(max_deliveryfile_size, self.replica_data['files'])
         output_name_prefix = self._create_file_name_prefix(reference)
         font = ImageFont.truetype('./font/Arial.ttf', 16)
         images = []
         n = 1
+        parts = []
+        count_images = 0
+        start = self._post_progress(iaid, 1)
+        if start:
```

Comment:
This if/else block is missing from the _create_zip def, should that do the same thing? If not, is there any point saving the outcome to the start variable in that def?

Reply:
This is reporting - back to the API and to CloudWatch. If this fails, I didn't think it should stop the process. Only log it in CloudWatch. Does this sound ok?

Reply:
I think so - the only thing I think is counterintuitive is that you do something different in the _create_pdf case than the _create_zip case. They should probably both do the same thing.

Reply:
Oh, yes, I missed that. I'll add the same zip function.

Reply:
I actually missed a bunch of code. I've now added all need for progress reporting. It could be done probably better...